_SPIKE_.md

```diff
@@ -0,0 +1,297 @@
+        categoryId: {type: 'number'},
+        category: {$ref: '#/definitions/CategoryWithLinks'},
+      },
+    },
+  },
+}
+```
+
+### Controller spec
+
+The last missing piece is integration with controller spec builder.
+
+At the moment, we use the following implementation of the controller method
+`find`:
+
+```ts
+class CategoryController {
+  // constructor with @inject() decorators
+
+  @get('/categories', {
+    responses: {
+      '200': {
+        description: 'Array of Category model instances',
+        content: {
+          'application/json': {
+            schema: {
+              type: 'array',
+              items: {'x-ts-type': Category},
+            },
+          },
+        },
+      },
+    },
+  })
+  async find(
+    @param.query.object('filter', getFilterSchemaFor(Category)) filter?: Filter,
+  ): Promise<TodoList[]> {
+    return await this.categoryRepository.find(filter);
+  }
+}
+```
+
+In my proposal, we replace `x-ts-type` extension with a call to
+`getModelSchemaRef`.
+
+```diff
+   schema: {
+     type: 'array',
+-    items: {'x-ts-type': Category},
++    items: getModelSchemaRef(Category, {includeRelations: true})
+   },
+```
+
+I particularly like how this solution makes it easy to use a different mechanism
+for generating model schema. Consider developers using TypeORM instead of our
+Juggler-based Repository for an example. If we implement my proposal, then it
+will be possible to implement a TypeORM extension for LoopBack that will provide
+a different implementation of `getModelSchemaRef` function, one that will use
+TypeORM metadata instead of juggler's LDL.
+
+Later on, we can explore different ways how to enable `includeRelations` flag
+via OpenAPI spec extensions. For example:
+
+```diff
+   schema: {
+     type: 'array',
+-    items: {'x-ts-type': Category},
++    items: {'x-ts-type': Category, 'x-include-relations': true},
+   },
+```
+
+## Follow-up stories
+
+1. Handle circular references when generating model JSON Schema
+
+2. Support `schema: {$ref, definitions}` in resolveControllerSpec
+
+3. Enhance `getJsonSchema` to describe navigational properties (introduce
+   `includeRelations` option).
+
+- Add a new `RelationDefinition` property: `targetsMany: boolean`
+- Implement support for `includeRelations` in `getJsonSchema` & co.
+
+4. Implement `getJsonSchemaRef` and `getModelSchemaRef` helpers
+
+5. Modify Repository `find*` method signatures to include links (navigational
+   properties) in the description of the return type
+
+- Add a new generic parameter `Links` to CRUD-related Repository interfaces and
+  implementations.
+- Modify the signature `find` and `findById` to return `T & Partial<Links>`
+  instead of `T`.
+
+6. Update `examples/todo-list` to leverage these new features:
+
+- Define `{Model}Links` interfaces
+- Update `{Model}Repository` implementations to use these new interfaces
+- Update repositories to include related models: overwrite `find` and `findById`
```

Comment:
Should we create a follow up story to replace our current poor man's way of retrieving related items?